packages/core/src/libraries/domain.ts

```diff
@@ -0,0 +1,109 @@
+import {
+  type CloudflareData,
+  type Domain,
+  type DomainDnsRecords,
+  DomainStatus,
+} from '@logto/schemas';
+import { conditional } from '@silverhand/essentials';
+
+import type Queries from '#src/tenants/Queries.js';
+import SystemContext from '#src/tenants/SystemContext.js';
+import assertThat from '#src/utils/assert-that.js';
+import { getCustomHostname, createCustomHostname } from '#src/utils/cloudflare/index.js';
+
+export type DomainLibrary = ReturnType<typeof createDomainLibrary>;
+
+const getDomainStatusFromCloudflareData = (data: CloudflareData): DomainStatus => {
+  if (data.status === 'active' && data.ssl.status === 'active') {
+    return DomainStatus.Active;
+  }
+
+  if (data.status === 'active' && data.ssl.status !== 'active') {
+    return DomainStatus.PendingSsl;
+  }
+
+  if (data.status !== 'active') {
+    return DomainStatus.PendingVerification;
+  }
+
+  return DomainStatus.Error;
+};
+
+export const createDomainLibrary = (queries: Queries) => {
+  const {
+    domains: { updateDomainById },
+  } = queries;
+
+  const syncDomainStatusFromCloudflareData = async (
+    domain: Domain,
+    cloudflareData: CloudflareData,
+    origin: string
+  ): Promise<Domain> => {
+    const status = getDomainStatusFromCloudflareData(cloudflareData);
+
+    const errorMessage = [
+      cloudflareData.verification_errors,
+      cloudflareData.ssl.validation_errors?.map(({ message }) => message),
+    ]
```

Comment:
Ah, I think you need to spread merge the two arrays this time, right?
```suggestion
    const errorMessage = [
      ...cloudflareData.verification_errors,
      ...cloudflareData.ssl.validation_errors?.map(({ message }) => message),
    ]
```